app/controllers/uploads_controller.rb

```diff
@@ -200,8 +217,18 @@ def generate_presigned_put
     ).performed!
 
     file_name = params.require(:file_name)
+    file_size = params.require(:file_size).to_i
     type = params.require(:type)
 
+    # we can pre-emptively check size for attachments, but not for images
+    # as they may be further reduced in size by UploadCreator (at this point
+    # they may have already been reduced in size by preprocessors)
+    if !FileHelper.is_supported_image?(file_name)
+      if file_size >= SiteSetting.max_attachment_size_kb.kilobytes
+        return render_json_error(I18n.t("upload.attachments.too_large", max_size_kb: SiteSetting.max_attachment_size_kb), status: 422)
+      end
+    end
```

Comment:
```suggestion
    if !FileHelper.is_supported_image?(file_name) && file_size >= SiteSetting.max_attachment_size_kb.kilobytes 
      return (
        render_json_error(
          I18n.t("upload.attachments.too_large", 
          max_size_kb: SiteSetting.max_attachment_size_kb
        ), 
        status: 422
      )
    end
```

Not sure if the nested is required here.